plugin/pxr/maya/lib/pxrUsdMayaGL/testenv/testProxyShapeDuplicatePerformance.py

```diff
@@ -0,0 +1,164 @@
+
+        statsOutput = '\n'.join(statsOutputLines)
+        perfStatsFilePath = '%s/perfStats.raw' % cls._testDir
+        with open(perfStatsFilePath, 'w') as perfStatsFile:
+            perfStatsFile.write(statsOutput)
+
+    def setUp(self):
+        cmds.file(new=True, force=True)
+
+        # To control where the rendered images are written, we force Maya to
+        # use the test directory as the workspace.
+        cmds.workspace(self._testDir, o=True)
+
+    @contextlib.contextmanager
+    def _ProfileScope(self, profileScopeName):
+        """
+        A context manager that measures the execution time between enter and
+        exit and stores the elapsed time in the class' metrics dictionary.
+        """
+        stopwatch = Tf.Stopwatch()
+
+        try:
+            stopwatch.Start()
+            yield
+        finally:
+            stopwatch.Stop()
+            elapsedTime = stopwatch.seconds
+            self._profileScopeMetrics[profileScopeName] = elapsedTime
+            Tf.Status("%s: %f" % (profileScopeName, elapsedTime))
+
+    def _RunLoadTest(self):
+        profileScopeName = '%s Assemblies Load Time' % self._testName
+
+        with self._ProfileScope(profileScopeName):
+            UsdMaya.LoadReferenceAssemblies()
+
+    def _RunDuplicateTest(self):
+        profileScopeName = '%s Assembly Duplicate' % self._testName
+
+        cmds.select('AssetRef_0_0_0')
+        cmds.refresh(force=True)
+
+        with self._ProfileScope(profileScopeName):
+            cmds.duplicate()
+            cmds.refresh(force=True)
+
+    def _WriteViewportImage(self, outputImageName, suffix):
+        # Make sure the hardware renderer is available
+        MAYA_RENDERER_NAME = 'mayaHardware2'
+        mayaRenderers = cmds.renderer(query=True, namesOfAvailableRenderers=True)
+        self.assertIn(MAYA_RENDERER_NAME, mayaRenderers)
+
+        # Make it the current renderer.
+        cmds.setAttr('defaultRenderGlobals.currentRenderer', MAYA_RENDERER_NAME,
+            type='string')
+        # Set the image format to PNG.
+        cmds.setAttr('defaultRenderGlobals.imageFormat', 32)
+        # Set the render mode to shaded and textured.
+        cmds.setAttr('hardwareRenderingGlobals.renderMode', 4)
+        # Specify the output image prefix. The path to it is built from the
+        # workspace directory.
+        cmds.setAttr('defaultRenderGlobals.imageFilePrefix',
+            '%s_%s' % (outputImageName, suffix),
+            type='string')
+        # Apply the viewer's color transform to the rendered image, otherwise
+        # it comes out too dark.
+        cmds.setAttr("defaultColorMgtGlobals.outputTransformEnabled", 1)
+
+        # Do the render.
+        cmds.ogsRender(camera=self._cameraName, currentFrame=True, width=960,
+            height=540)
+
+    def testPerfGridOfCubeModelDuplicate(self):
+        """
+        Tests speed of duplicating a model reference assembly/proxy when many
```

Comment:
How many are `many`?

Reply:
Yeah, sorry, we've been intentionally leaving the comments vague in terms of numbers to try to avoid them going stale in case we need to scale up the tests to keep pace with development.

This test is scaled back a bit from usual production workloads to make for a reasonable unit test, but at the moment it's a 5x5x5 cubic grid for a total of 125 of `pxrUsdProxyShape`. Each one of those proxy shapes references a USD stage containing 10x10x10 Mesh prims. Because of our use (read: abuse) of the Hydra API and our high render task counts, we're tripping over some O(N^2) behavior during Hydra sync where we visit every rprim in the index for every render task:
https://github.com/PixarAnimationStudios/USD/blob/6af4d8afd02682ee4365d1b7d60683e95acf12ac/pxr/imaging/hd/dirtyList.cpp#L107

Internally, this particular test takes about 4 seconds to duplicate one of those proxy shapes. With production assets, it's not uncommon to have 1000+ proxy shapes in a medium to large set, and artists are sometimes seeing duplicates take as much as 15-20 seconds.

We're working to address that now, so I'll likely have some follow-up PRs around this for `pxrUsdMayaGL`.

Reply:
Thank you for providing extra info. Have you tried comparing performance with VP2RenderDelegate? FYI. We are actively working on improving instancing and consolidation performance with subscene override (an API used by VP2RenderDelegate). It would be good to know if there is something else we should be measuring when doing this work.

Reply:
We haven't yet tested this particular issue with the VP2.0 render delegate, but we'll hopefully be looking at that soon. Once I'm able to devote some time to  that, one of the first things will probably be to make sure that all of these pxrUsdMayaGL tests can be run externally so that we're measuring the same things for both viewport integrations.